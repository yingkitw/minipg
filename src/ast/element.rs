//! Grammar element AST nodes.

use serde::{Deserialize, Serialize};

/// Lexer command (e.g., skip, channel, mode, type)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum LexerCommand {
    /// Skip this token
    Skip,
    /// Send to a channel
    Channel(String),
    /// Switch to a mode
    Mode(String),
    /// Change token type
    Type(String),
    /// Push mode
    PushMode(String),
    /// Pop mode
    PopMode,
    /// More (continue current token)
    More,
}

/// An alternative in a rule (sequence of elements).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Alternative {
    pub elements: Vec<Element>,
    pub label: Option<String>,
    pub lexer_command: Option<LexerCommand>,
}

impl Alternative {
    pub fn new() -> Self {
        Self {
            elements: Vec::new(),
            label: None,
            lexer_command: None,
        }
    }

    pub fn with_label(mut self, label: String) -> Self {
        self.label = Some(label);
        self
    }

    pub fn with_lexer_command(mut self, command: LexerCommand) -> Self {
        self.lexer_command = Some(command);
        self
    }

    pub fn add_element(&mut self, element: Element) {
        self.elements.push(element);
    }

    pub fn set_lexer_command(&mut self, command: LexerCommand) {
        self.lexer_command = Some(command);
    }
}

impl Default for Alternative {
    fn default() -> Self {
        Self::new()
    }
}

/// Grammar element (terminal, non-terminal, etc.).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Element {
    /// Reference to another rule
    RuleRef {
        name: String,
        label: Option<String>,
    },
    /// Terminal token
    Terminal {
        value: String,
        label: Option<String>,
    },
    /// String literal
    StringLiteral {
        value: String,
        label: Option<String>,
    },
    /// Character range (for lexer rules)
    CharRange {
        start: char,
        end: char,
    },
    /// Character class (e.g., [a-z0-9], ~["\r\n])
    CharClass {
        negated: bool,
        ranges: Vec<(char, char)>, // Vec of (start, end) - single char is (c, c)
    },
    /// Optional element (?)
    Optional {
        element: Box<Element>,
        greedy: bool, // true for ?, false for ??
    },
    /// Zero or more (*)
    ZeroOrMore {
        element: Box<Element>,
        greedy: bool, // true for *, false for *?
    },
    /// One or more (+)
    OneOrMore {
        element: Box<Element>,
        greedy: bool, // true for +, false for +?
    },
    /// Grouped elements
    Group {
        alternatives: Vec<Alternative>,
    },
    /// Negation (~)
    Not {
        element: Box<Element>,
    },
    /// Any character (.)
    Wildcard,
    /// End of file
    Eof,
    /// Embedded action code
    Action {
        code: String,
        language: Option<String>,
    },
    /// Semantic predicate
    Predicate {
        code: String,
        language: Option<String>,
    },
}

impl Element {
    pub fn rule_ref(name: String) -> Self {
        Element::RuleRef { name, label: None }
    }

    pub fn terminal(value: String) -> Self {
        Element::Terminal { value, label: None }
    }

    pub fn string_literal(value: String) -> Self {
        Element::StringLiteral { value, label: None }
    }

    pub fn optional(element: Element) -> Self {
        Element::Optional {
            element: Box::new(element),
            greedy: true,
        }
    }

    pub fn optional_non_greedy(element: Element) -> Self {
        Element::Optional {
            element: Box::new(element),
            greedy: false,
        }
    }

    pub fn zero_or_more(element: Element) -> Self {
        Element::ZeroOrMore {
            element: Box::new(element),
            greedy: true,
        }
    }

    pub fn zero_or_more_non_greedy(element: Element) -> Self {
        Element::ZeroOrMore {
            element: Box::new(element),
            greedy: false,
        }
    }

    pub fn one_or_more(element: Element) -> Self {
        Element::OneOrMore {
            element: Box::new(element),
            greedy: true,
        }
    }

    pub fn one_or_more_non_greedy(element: Element) -> Self {
        Element::OneOrMore {
            element: Box::new(element),
            greedy: false,
        }
    }

    pub fn with_label(self, label: String) -> Self {
        match self {
            Element::RuleRef { name, .. } => Element::RuleRef {
                name,
                label: Some(label),
            },
            Element::Terminal { value, .. } => Element::Terminal {
                value,
                label: Some(label),
            },
            Element::StringLiteral { value, .. } => Element::StringLiteral {
                value,
                label: Some(label),
            },
            other => other,
        }
    }

    pub fn action(code: String) -> Self {
        Element::Action {
            code,
            language: None,
        }
    }

    pub fn action_with_language(code: String, language: String) -> Self {
        Element::Action {
            code,
            language: Some(language),
        }
    }

    pub fn predicate(code: String) -> Self {
        Element::Predicate {
            code,
            language: None,
        }
    }

    pub fn predicate_with_language(code: String, language: String) -> Self {
        Element::Predicate {
            code,
            language: Some(language),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_alternative_new() {
        let alt = Alternative::new();
        assert_eq!(alt.elements.len(), 0);
        assert_eq!(alt.label, None);
    }

    #[test]
    fn test_alternative_with_label() {
        let alt = Alternative::new().with_label("myLabel".to_string());
        assert_eq!(alt.label, Some("myLabel".to_string()));
    }

    #[test]
    fn test_alternative_add_element() {
        let mut alt = Alternative::new();
        alt.add_element(Element::Terminal {
            value: "ID".to_string(),
            label: None,
        });
        assert_eq!(alt.elements.len(), 1);
    }

    #[test]
    fn test_element_rule_ref() {
        let elem = Element::RuleRef {
            name: "expr".to_string(),
            label: Some("e".to_string()),
        };
        match elem {
            Element::RuleRef { name, label } => {
                assert_eq!(name, "expr");
                assert_eq!(label, Some("e".to_string()));
            }
            _ => panic!("Expected RuleRef"),
        }
    }

    #[test]
    fn test_element_terminal() {
        let elem = Element::Terminal {
            value: "ID".to_string(),
            label: None,
        };
        match elem {
            Element::Terminal { value, label } => {
                assert_eq!(value, "ID");
                assert_eq!(label, None);
            }
            _ => panic!("Expected Terminal"),
        }
    }

    #[test]
    fn test_element_action() {
        let elem = Element::action("System.out.println();".to_string());
        match elem {
            Element::Action { code, language } => {
                assert_eq!(code, "System.out.println();");
                assert_eq!(language, None);
            }
            _ => panic!("Expected Action"),
        }
    }

    #[test]
    fn test_element_predicate() {
        let elem = Element::predicate("x > 0".to_string());
        match elem {
            Element::Predicate { code, language } => {
                assert_eq!(code, "x > 0");
                assert_eq!(language, None);
            }
            _ => panic!("Expected Predicate"),
        }
    }

    #[test]
    fn test_char_class_simple() {
        let elem = Element::CharClass {
            negated: false,
            ranges: vec![('a', 'z')],
        };
        match elem {
            Element::CharClass { negated, ranges } => {
                assert!(!negated);
                assert_eq!(ranges.len(), 1);
                assert_eq!(ranges[0], ('a', 'z'));
            }
            _ => panic!("Expected CharClass"),
        }
    }

    #[test]
    fn test_char_class_negated() {
        let elem = Element::CharClass {
            negated: true,
            ranges: vec![('"', '"'), ('\\', '\\'), ('\u{0000}', '\u{001F}')],
        };
        match elem {
            Element::CharClass { negated, ranges } => {
                assert!(negated);
                assert_eq!(ranges.len(), 3);
            }
            _ => panic!("Expected CharClass"),
        }
    }

    #[test]
    fn test_char_class_multiple_ranges() {
        let elem = Element::CharClass {
            negated: false,
            ranges: vec![('a', 'z'), ('A', 'Z'), ('0', '9')],
        };
        match elem {
            Element::CharClass { negated, ranges } => {
                assert!(!negated);
                assert_eq!(ranges.len(), 3);
            }
            _ => panic!("Expected CharClass"),
        }
    }
}
