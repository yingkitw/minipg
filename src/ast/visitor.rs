//! Visitor pattern for AST traversal.

use super::{Alternative, Element, Grammar, Rule};

/// Immutable visitor for AST traversal.
pub trait AstVisitor {
    fn visit_grammar(&mut self, grammar: &Grammar) {
        self.walk_grammar(grammar);
    }

    fn visit_rule(&mut self, rule: &Rule) {
        self.walk_rule(rule);
    }

    fn visit_alternative(&mut self, alternative: &Alternative) {
        self.walk_alternative(alternative);
    }

    fn visit_element(&mut self, element: &Element) {
        self.walk_element(element);
    }

    fn walk_grammar(&mut self, grammar: &Grammar) {
        for rule in &grammar.rules {
            self.visit_rule(rule);
        }
    }

    fn walk_rule(&mut self, rule: &Rule) {
        for alternative in &rule.alternatives {
            self.visit_alternative(alternative);
        }
    }

    fn walk_alternative(&mut self, alternative: &Alternative) {
        for element in &alternative.elements {
            self.visit_element(element);
        }
    }

    fn walk_element(&mut self, element: &Element) {
        match element {
            Element::Optional { element } => self.visit_element(element),
            Element::ZeroOrMore { element } => self.visit_element(element),
            Element::OneOrMore { element } => self.visit_element(element),
            Element::Not { element } => self.visit_element(element),
            Element::Group { alternatives } => {
                for alt in alternatives {
                    self.visit_alternative(alt);
                }
            }
            _ => {}
        }
    }
}

/// Mutable visitor for AST transformation.
pub trait AstVisitorMut {
    fn visit_grammar_mut(&mut self, grammar: &mut Grammar) {
        self.walk_grammar_mut(grammar);
    }

    fn visit_rule_mut(&mut self, rule: &mut Rule) {
        self.walk_rule_mut(rule);
    }

    fn visit_alternative_mut(&mut self, alternative: &mut Alternative) {
        self.walk_alternative_mut(alternative);
    }

    fn visit_element_mut(&mut self, element: &mut Element) {
        self.walk_element_mut(element);
    }

    fn walk_grammar_mut(&mut self, grammar: &mut Grammar) {
        for rule in &mut grammar.rules {
            self.visit_rule_mut(rule);
        }
    }

    fn walk_rule_mut(&mut self, rule: &mut Rule) {
        for alternative in &mut rule.alternatives {
            self.visit_alternative_mut(alternative);
        }
    }

    fn walk_alternative_mut(&mut self, alternative: &mut Alternative) {
        for element in &mut alternative.elements {
            self.visit_element_mut(element);
        }
    }

    fn walk_element_mut(&mut self, element: &mut Element) {
        match element {
            Element::Optional { element } => self.visit_element_mut(element),
            Element::ZeroOrMore { element } => self.visit_element_mut(element),
            Element::OneOrMore { element } => self.visit_element_mut(element),
            Element::Not { element } => self.visit_element_mut(element),
            Element::Group { alternatives } => {
                for alt in alternatives {
                    self.visit_alternative_mut(alt);
                }
            }
            _ => {}
        }
    }
}
