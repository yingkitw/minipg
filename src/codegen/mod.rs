//! Code generation for parser generators.

pub mod rust;
pub mod python;
pub mod javascript;
pub mod typescript;
pub mod go;
pub mod c;
pub mod cpp;
pub mod java;
pub mod template;
pub mod visitor_gen;
pub mod dfa;
pub mod lookup_table;
pub mod modes;
pub mod actions;

pub use rust::RustCodeGenerator;
pub use python::PythonCodeGenerator;
pub use javascript::JavaScriptCodeGenerator;
pub use typescript::TypeScriptCodeGenerator;
pub use go::GoCodeGenerator;
pub use c::CCodeGenerator;
pub use cpp::CppCodeGenerator;
pub use java::JavaCodeGenerator;

use crate::analysis::AnalysisResult;
use crate::core::{types::CodeGenConfig, CodeGenerator as CodeGeneratorTrait, Result};

/// Main code generator dispatcher.
pub struct CodeGenerator {
    config: CodeGenConfig,
}

impl CodeGenerator {
    pub fn new(config: CodeGenConfig) -> Self {
        Self { config }
    }

    pub fn generate_from_analysis(&self, analysis: &AnalysisResult) -> Result<String> {
        match self.config.target_language.as_str() {
            "rust" => {
                let generator = RustCodeGenerator::new();
                generator.generate(&analysis.grammar, &self.config)
            }
            "python" => {
                let generator = PythonCodeGenerator::new();
                generator.generate(&analysis.grammar, &self.config)
            }
            "javascript" | "js" => {
                let generator = JavaScriptCodeGenerator::new();
                generator.generate(&analysis.grammar, &self.config)
            }
            "typescript" | "ts" => {
                let generator = TypeScriptCodeGenerator::new();
                generator.generate(&analysis.grammar, &self.config)
            }
            "go" => {
                let generator = GoCodeGenerator::new();
                generator.generate(&analysis.grammar, &self.config)
            }
            "c" => {
                let generator = CCodeGenerator::new();
                generator.generate(&analysis.grammar, &self.config)
            }
            "cpp" | "c++" => {
                let generator = CppCodeGenerator::new();
                generator.generate(&analysis.grammar, &self.config)
            }
            "java" => {
                let generator = JavaCodeGenerator::new();
                generator.generate(&analysis.grammar, &self.config)
            }
            _ => Err(crate::core::Error::codegen(format!(
                "unsupported target language: {}",
                self.config.target_language
            ))),
        }
    }
}
