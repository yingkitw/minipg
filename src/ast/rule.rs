//! Rule AST node definitions.

use super::element::Alternative;
use serde::{Deserialize, Serialize};

/// Type of rule.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum RuleType {
    Parser,
    Lexer,
}

/// Rule argument definition
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuleArg {
    pub name: String,
    pub arg_type: Option<String>,
}

/// Rule return value definition
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuleReturn {
    pub name: String,
    pub return_type: Option<String>,
}

/// Rule local variable definition
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuleLocal {
    pub name: String,
    pub local_type: Option<String>,
}

/// A grammar rule.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Rule {
    pub name: String,
    pub rule_type: RuleType,
    pub alternatives: Vec<Alternative>,
    pub is_fragment: bool,
    /// Rule arguments: rule[int x, String name]
    pub arguments: Vec<RuleArg>,
    /// Return values: returns [Type value]
    pub returns: Vec<RuleReturn>,
    /// Local variables: locals [Type var]
    pub locals: Vec<RuleLocal>,
}

impl Rule {
    pub fn new(name: String, rule_type: RuleType) -> Self {
        Self {
            name,
            rule_type,
            alternatives: Vec::new(),
            is_fragment: false,
            arguments: Vec::new(),
            returns: Vec::new(),
            locals: Vec::new(),
        }
    }
    
    pub fn add_argument(&mut self, name: String, arg_type: Option<String>) {
        self.arguments.push(RuleArg { name, arg_type });
    }
    
    pub fn add_return(&mut self, name: String, return_type: Option<String>) {
        self.returns.push(RuleReturn { name, return_type });
    }
    
    pub fn add_local(&mut self, name: String, local_type: Option<String>) {
        self.locals.push(RuleLocal { name, local_type });
    }

    pub fn parser_rule(name: String) -> Self {
        Self::new(name, RuleType::Parser)
    }

    pub fn lexer_rule(name: String) -> Self {
        Self::new(name, RuleType::Lexer)
    }

    pub fn add_alternative(&mut self, alternative: Alternative) {
        self.alternatives.push(alternative);
    }

    pub fn is_lexer_rule(&self) -> bool {
        self.rule_type == RuleType::Lexer
    }

    pub fn is_parser_rule(&self) -> bool {
        self.rule_type == RuleType::Parser
    }

    pub fn set_fragment(&mut self, is_fragment: bool) {
        self.is_fragment = is_fragment;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_rule_with_arguments() {
        let mut rule = Rule::parser_rule("test".to_string());
        rule.add_argument("x".to_string(), Some("int".to_string()));
        rule.add_argument("name".to_string(), Some("String".to_string()));
        
        assert_eq!(rule.arguments.len(), 2);
        assert_eq!(rule.arguments[0].name, "x");
        assert_eq!(rule.arguments[0].arg_type, Some("int".to_string()));
        assert_eq!(rule.arguments[1].name, "name");
        assert_eq!(rule.arguments[1].arg_type, Some("String".to_string()));
    }

    #[test]
    fn test_rule_with_returns() {
        let mut rule = Rule::parser_rule("test".to_string());
        rule.add_return("result".to_string(), Some("Value".to_string()));
        
        assert_eq!(rule.returns.len(), 1);
        assert_eq!(rule.returns[0].name, "result");
        assert_eq!(rule.returns[0].return_type, Some("Value".to_string()));
    }

    #[test]
    fn test_rule_with_locals() {
        let mut rule = Rule::parser_rule("test".to_string());
        rule.add_local("temp".to_string(), Some("int".to_string()));
        rule.add_local("buffer".to_string(), Some("String".to_string()));
        
        assert_eq!(rule.locals.len(), 2);
        assert_eq!(rule.locals[0].name, "temp");
        assert_eq!(rule.locals[1].name, "buffer");
    }

    #[test]
    fn test_rule_with_all_features() {
        let mut rule = Rule::parser_rule("complex".to_string());
        rule.add_argument("input".to_string(), Some("String".to_string()));
        rule.add_return("output".to_string(), Some("Result".to_string()));
        rule.add_local("state".to_string(), Some("int".to_string()));
        
        assert_eq!(rule.arguments.len(), 1);
        assert_eq!(rule.returns.len(), 1);
        assert_eq!(rule.locals.len(), 1);
        assert!(rule.is_parser_rule());
    }

    #[test]
    fn test_rule_arg_without_type() {
        let mut rule = Rule::parser_rule("test".to_string());
        rule.add_argument("x".to_string(), None);
        
        assert_eq!(rule.arguments.len(), 1);
        assert_eq!(rule.arguments[0].name, "x");
        assert_eq!(rule.arguments[0].arg_type, None);
    }
}
