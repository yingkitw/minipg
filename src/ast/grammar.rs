//! Grammar AST node definitions.

use super::rule::Rule;
use crate::core::types::GrammarType;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Root node of a grammar AST.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Grammar {
    pub name: String,
    pub grammar_type: GrammarType,
    pub options: HashMap<String, String>,
    pub rules: Vec<Rule>,
    pub imports: Vec<String>,
    /// Named actions like @header, @members, etc.
    pub named_actions: HashMap<String, String>,
    /// Lexer modes: mode_name -> rules in that mode
    pub lexer_modes: HashMap<String, Vec<String>>,
    /// Channel names used in the grammar
    pub channels: std::collections::HashSet<String>,
}

impl Grammar {
    pub fn new(name: String, grammar_type: GrammarType) -> Self {
        Self {
            name,
            grammar_type,
            options: HashMap::new(),
            rules: Vec::new(),
            imports: Vec::new(),
            named_actions: HashMap::new(),
            lexer_modes: HashMap::new(),
            channels: std::collections::HashSet::new(),
        }
    }

    pub fn add_rule(&mut self, rule: Rule) {
        self.rules.push(rule);
    }

    pub fn add_option(&mut self, key: String, value: String) {
        self.options.insert(key, value);
    }

    pub fn add_import(&mut self, import: String) {
        self.imports.push(import);
    }

    pub fn add_named_action(&mut self, name: String, code: String) {
        self.named_actions.insert(name, code);
    }

    pub fn get_rule(&self, name: &str) -> Option<&Rule> {
        self.rules.iter().find(|r| r.name == name)
    }

    pub fn lexer_rules(&self) -> impl Iterator<Item = &Rule> {
        self.rules.iter().filter(|r| r.is_lexer_rule())
    }

    pub fn parser_rules(&self) -> impl Iterator<Item = &Rule> {
        self.rules.iter().filter(|r| r.is_parser_rule())
    }

    pub fn add_lexer_mode(&mut self, mode_name: String, rule_names: Vec<String>) {
        self.lexer_modes.insert(mode_name, rule_names);
    }

    pub fn add_channel(&mut self, channel_name: String) {
        self.channels.insert(channel_name);
    }

    pub fn has_modes(&self) -> bool {
        !self.lexer_modes.is_empty()
    }

    pub fn has_channels(&self) -> bool {
        !self.channels.is_empty()
    }
}

/// Generic grammar node for AST traversal.
#[derive(Debug, Clone)]
pub enum GrammarNode {
    Grammar(Grammar),
    Rule(Rule),
}
