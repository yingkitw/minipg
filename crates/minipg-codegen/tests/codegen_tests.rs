//! Code generation tests.

use minipg::ast::{Alternative, Element, Grammar, Rule};
use minipg::codegen::RustCodeGenerator;
use minipg::core::{types::{CodeGenConfig, GrammarType}, CodeGenerator};

#[test]
fn test_rust_codegen_simple() {
    let mut grammar = Grammar::new("Calculator".to_string(), GrammarType::Parser);
    
    let mut expr_rule = Rule::parser_rule("expr".to_string());
    let mut alt = Alternative::new();
    alt.add_element(Element::rule_ref("term".to_string()));
    expr_rule.add_alternative(alt);
    grammar.add_rule(expr_rule);
    
    let mut term_rule = Rule::parser_rule("term".to_string());
    let mut alt = Alternative::new();
    alt.add_element(Element::rule_ref("NUMBER".to_string()));
    term_rule.add_alternative(alt);
    grammar.add_rule(term_rule);
    
    let mut number_rule = Rule::lexer_rule("NUMBER".to_string());
    let mut alt = Alternative::new();
    alt.add_element(Element::one_or_more(Element::CharRange {
        start: '0',
        end: '9',
    }));
    number_rule.add_alternative(alt);
    grammar.add_rule(number_rule);

    let generator = RustCodeGenerator::new();
    let config = CodeGenConfig::default();
    let result = generator.generate(&grammar, &config);
    
    assert!(result.is_ok());
    let code = result.unwrap();
    
    insta::assert_snapshot!(code);
}
