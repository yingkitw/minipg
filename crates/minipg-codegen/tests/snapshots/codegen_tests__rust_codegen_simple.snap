---
source: crates/minipg-codegen/tests/codegen_tests.rs
expression: code
---
// Generated parser for Calculator grammar
// DO NOT EDIT - This file is automatically generated

use std::fmt;

/// Parse error with context information.
#[derive(Debug, Clone, PartialEq)]
pub struct ParseError {
    pub message: String,
    pub position: usize,
    pub expected: Vec<String>,
    pub found: Option<String>,
}

impl ParseError {
    pub fn new(message: String, position: usize) -> Self {
        Self {
            message,
            position,
            expected: Vec::new(),
            found: None,
        }
    }

    pub fn with_expected(mut self, expected: Vec<String>) -> Self {
        self.expected = expected;
        self
    }

    pub fn with_found(mut self, found: String) -> Self {
        self.found = Some(found);
        self
    }
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Parse error at position {}: {}", self.position, self.message)?;
        if !self.expected.is_empty() {
            write!(f, " (expected: {})", self.expected.join(", "))?;
        }
        if let Some(ref found) = self.found {
            write!(f, " (found: {})", found)?;
        }
        Ok(())
    }
}

impl std::error::Error for ParseError {}

/// Token with position information.
#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub text: String,
    pub position: usize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    NUMBER,
    Eof,
}

/// AST node types.
#[derive(Debug, Clone)]
pub enum AstNode {
    Expr(Box<Expr>),
    Term(Box<Term>),
}

/// Listener trait for AST events.
pub trait Listener {
    fn enter_expr(&mut self, _node: &Expr) {}
    fn exit_expr(&mut self, _node: &Expr) {}
    fn enter_term(&mut self, _node: &Term) {}
    fn exit_term(&mut self, _node: &Term) {}
}

/// Lexer for Calculator grammar.
/// 
/// This lexer uses an optimized DFA (Deterministic Finite Automaton)
/// generated at compile time for efficient tokenization.
#[derive(Debug)]
pub struct CalculatorLexer {
    input: Vec<char>,
    position: usize,
}

impl CalculatorLexer {
    /// Create a new lexer from input string.
    #[inline]
    pub fn new(input: &str) -> Self {
        Self {
            input: input.chars().collect(),
            position: 0,
        }
    }

    /// Get the next token from the input.
    /// 
    /// Returns Ok(Token) on success, or Err(ParseError) if tokenization fails.
    pub fn next_token(&mut self) -> Result<Token, ParseError> {
        // Skip whitespace
        self.skip_whitespace();

        let start_pos = self.position;

        // Check for EOF
        if self.position >= self.input.len() {
            return Ok(Token {
                kind: TokenKind::Eof,
                text: String::new(),
                position: start_pos,
            });
        }

        // Use DFA for tokenization
        match self.next_token_dfa() {
            Some(mut token) => {
                token.position = start_pos;
                Ok(token)
            }
            None => {
                // Error recovery: skip invalid character and try again
                let invalid_char = self.input[self.position];
                self.position += 1;
                Err(ParseError::new(
                    format!("Unexpected character: '{}'", invalid_char),
                    start_pos,
                ).with_found(invalid_char.to_string()))
            }
        }
    }

    /// Tokenize all input and collect errors.
    /// 
    /// Returns all successfully parsed tokens and a list of errors encountered.
    pub fn tokenize_all(&mut self) -> (Vec<Token>, Vec<ParseError>) {
        let mut tokens = Vec::new();
        let mut errors = Vec::new();

        loop {
            match self.next_token() {
                Ok(token) => {
                    let is_eof = token.kind == TokenKind::Eof;
                    tokens.push(token);
                    if is_eof {
                        break;
                    }
                }
                Err(err) => {
                    errors.push(err);
                    // Continue tokenizing after error
                    if self.position >= self.input.len() {
                        break;
                    }
                }
            }
        }

        (tokens, errors)
    }

    #[inline(always)]
    fn skip_whitespace(&mut self) {
        while self.position < self.input.len() {
            match self.input[self.position] {
                ' ' | '\t' | '\r' | '\n' => self.position += 1,
                _ => break,
            }
        }
    }

    /// Character class lookup table.
    /// 
    /// Maps each character to its character class ID for efficient matching.
    /// This table is generated at compile time and stored as a const array.
    const CHAR_CLASS_TABLE: [u8; 256] = [
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0x00-0x0F
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0x10-0x1F
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0x20-0x2F
          0,   1,   2,   3,   4,   5,   6,   7,   8,   9, 255, 255, 255, 255, 255, 255,  // 0x30-0x3F
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0x40-0x4F
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0x50-0x5F
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0x60-0x6F
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0x70-0x7F
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0x80-0x8F
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0x90-0x9F
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0xA0-0xAF
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0xB0-0xBF
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0xC0-0xCF
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0xD0-0xDF
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,  // 0xE0-0xEF
        255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255 // 0xF0-0xFF

    ];

    /// Get the character class ID for a given character.
    /// 
    /// Returns 255 for characters not in any class.
    #[inline]
    fn get_char_class(ch: char) -> u8 {
        if (ch as u32) < 256 {
            Self::CHAR_CLASS_TABLE[ch as usize]
        } else {
            255 // Unknown class for non-ASCII
        }
    }

    /// Token type lookup table.
    /// 
    /// Maps token name strings to TokenKind for efficient conversion.
    #[inline]
    fn token_name_to_kind(name: &str) -> TokenKind {
        match name {
            "NUMBER" => TokenKind::NUMBER,
            _ => TokenKind::Eof,
        }
    }

    /// Match character using lookup table.
    /// 
    /// This is faster than multiple if/else or match statements.
    #[inline]
    fn match_char_fast(&self, ch: char, expected_class: u8) -> bool {
        Self::get_char_class(ch) == expected_class
    }

    /// Check if character is in a range using lookup table.
    #[inline]
    fn is_in_range(&self, ch: char, start_class: u8, end_class: u8) -> bool {
        let class = Self::get_char_class(ch);
        class >= start_class && class <= end_class
    }

    fn next_token_dfa(&mut self) -> Option<Token> {
        let mut state = 0;
        let mut token_start = self.position;
        let mut last_accepting: Option<(usize, &str)> = None;

        loop {
            // Check if current state is accepting
            match state {
                0 => last_accepting = Some((self.position, "NUMBER")),
                _ => {}
            }

            // Get next character
            let ch = match self.input.get(self.position) {
                Some(&c) => c,
                None => break,
            };

            // Transition to next state
            state = match (state, ch) {
                _ => break, // No valid transition
            };

            self.position += 1;
        }

        // Return token if we found an accepting state
        if let Some((end_pos, token_name)) = last_accepting {
            let text: String = self.input[token_start..end_pos].iter().collect();
            Some(Token {
                position: token_start,
                kind: match token_name {
                    "NUMBER" => TokenKind::NUMBER,
                    _ => TokenKind::Eof,
                },
                text,
            })
        } else {
            None
        }
    }
    // Lookup table stats: 10 chars, 10 classes, 256 bytes
}

/// Parser for Calculator grammar.
#[derive(Debug)]
pub struct CalculatorParser {
    tokens: Vec<Token>,
    position: usize,
}

impl CalculatorParser {
    #[inline]
    pub fn new(tokens: Vec<Token>) -> Self {
        Self { tokens, position: 0 }
    }

    /// Parse expr rule.
    pub fn parse_expr(&mut self) -> Result<AstNode> {
        // TODO: Implement rule parsing
        unimplemented!()
    }

    /// Parse term rule.
    pub fn parse_term(&mut self) -> Result<AstNode> {
        // TODO: Implement rule parsing
        unimplemented!()
    }

}
